src/libraries/System.Net.Security/src/System/Net/Security/NegotiateStream.cs

```diff
@@ -883,7 +883,16 @@ private async Task ReceiveBlobAsync<TIOAdapter>(CancellationToken cancellationTo
 
             if (_framer.ReadHeader.MessageId == FrameHeader.HandshakeDoneId)
             {
-                _remoteOk = true;
+                if (message.Length > 0)
+                {
+                    Debug.Assert(_context != null);
+                    _context.GetOutgoingBlob(message, out NegotiateAuthenticationStatusCode statusCode);
+                    _remoteOk = statusCode is NegotiateAuthenticationStatusCode.Completed;
+                }
```

Comment:
Review note:

This was likely broken for quite a long time and went unnoticed. Windows clients send raw NTLM instead of SPNEGO if NTLM is the only protocol available. Linux/macOS GSSAPI libraries are aware of this quirk and handle it too. In that case the last handshake message will be empty.

Skipping the processing of the last handshake message is mostly harmless. In SPNEGO it would contain the final message integrity check. Skipping this check could enable some man-in-the-middle attacks, at least in theory. In case of managed NTLM/SPNEGO, however, we reset the NTLM keys only after this very last message is processed. If this is skipped the client and server will have mismatched encryption keys. This is how the bug was discovered.

In theory we can align with GSSAPI by calling `(_mechanism as ManagedNtlmNegotiateAuthenticationPal)?.ResetKeys();` here:
https://github.com/dotnet/runtime/blob/1d2b1ae07edbd769380a8d20b30c3e9c18e2f649/src/libraries/System.Net.Security/src/System/Net/NegotiateAuthenticationPal.ManagedSpnego.cs#L350-L351
We would reset the key both after `GetMIC` and `VerifyMIC` instead of doing it only after `VerifyMIC`. It may not be desirable to do so though; when the API is used correctly and securely this is not an issue.